cpmpy/solvers/gurobi.py

```diff
@@ -0,0 +1,376 @@
+
+            `minimize()` can be called multiple times, only the last one is used
+
+            (technical side note: any constraints created during conversion of the objective
+            are premanently posted to the solver)
+        """
+        from gurobipy import GRB
+
+        # make objective function non-nested
+        (flat_obj, flat_cons) = (flatten_objective(expr))
+        self += linearize_constraint(flat_cons)  # add potentially created constraints
+
+        # make objective function or variable and post
+        obj = self._make_numexpr(flat_obj)
+        self.gbi_model.setObjective(obj, sense=GRB.MINIMIZE)
+
+    def maximize(self, expr):
+        """
+            Maximize the given objective function
+
+            `maximize()` can be called multiple times, only the last one is used
+
+            (technical side note: any constraints created during conversion of the objective
+            are premanently posted to the solver)
+        """
+        from gurobipy import GRB
+
+        # make objective function non-nested
+        (flat_obj, flat_cons) = (flatten_objective(expr))
+        self += linearize_constraint(flat_cons)  # add potentially created constraints
+
+        # make objective function or variable and post
+        obj = self._make_numexpr(flat_obj)
+        self.gbi_model.setObjective(obj, sense=GRB.MAXIMIZE)
+
+    def _make_numexpr(self, cpm_expr):
+        """
+            Turns a numeric CPMpy 'flat' expression into a solver-specific
+            numeric expression
+
+            Used especially to post an expression as objective function
+        """
+        import gurobipy as gp
+
+        if is_num(cpm_expr):
+            return cpm_expr
+
+        # decision variables, check in varmap
+        if isinstance(cpm_expr, _NumVarImpl):  # _BoolVarImpl is subclass of _NumVarImpl
+            return self.solver_var(cpm_expr)
+
+        # sum
+        if cpm_expr.name == "sum":
+            return gp.quicksum([self.solver_var(var) for var in cpm_expr.args])
+
+        if cpm_expr.name == "wsum":
+            weights, cpm_vars = cpm_expr.args
+            gbi_vars = [self.solver_var(var) for var in cpm_vars]
+            return gp.LinExpr(weights, gbi_vars)
+
+        raise NotImplementedError("gurobi: Not a know supported numexpr {}".format(cpm_expr))
+
+    def _post_constraint(self, cpm_expr):
+        """
+            Post a primitive CPMpy constraint to the native solver API
+
+            What 'primitive' means depends on the solver capabilities,
+            more specifically on the transformations applied in `__add__()`
+
+            Solvers do not need to support all constraints.
+        """
+
+        # Base case: Boolean variable
+        if isinstance(cpm_expr, _BoolVarImpl):
+            return self.gbi_model.addLConstr(self.solver_var(cpm_expr), ">", 1)
+
+
+        # Operators: base (bool), lhs=numexpr, lhs|rhs=boolexpr (reified ->)
+        elif isinstance(cpm_expr, Operator):
```

Comment:
I'm not sure you should keep this in just to throw an error... you can assume that the input to this function have been through your transformations and hence only of the expressions you want (and if not throw a general NotImplemented or Exception("unexpected...") at the end of this function...)